Use the Xunit assert library instead of CoreCLRTestLibrary for most asserts

jkoritzinsky opened the pull request:
Asserts not provided by Xunit are provided by an AssertExtensions class.

Tests that reference System.Private.Corelib directly will use a polyfill implementation based off the old CoreCLRTestLibrary asserts.

All assert methods provided by CoreCLRTestLibrary have been changed to follow Xunit conventions (this is why the change is so noisy).

This contribues to our move towards more xunit-style tests in the src/tests tree, just with a custom runner (as part of the test merging initiative).

cc: @trylek for review

cc: @elinor-fung and @AaronRobinsonMSFT as the Interop team owns or has worked on most of the tests that are modified by this change.

jkoritzinsky committed:
Use the Xunit assert library instead of CoreCLRTestLibrary for most asserts.

Asserts not provided by Xunit are provided by an AssertExtensions class.

Tests that reference System.Private.Corelib directly will use a polyfill implementation based off the old CoreCLRTestLibrary asserts.

All assert methods provided by CoreCLRTestLibrary have been changed to follow Xunit conventions.

AaronRobinsonMSFT commented on src/tests/Common/Assert.cs:
```diff
@@ -0,0 +1,240 @@
+        [Obsolete("Did you mean to call Assert.Equal()")]
+        public static new bool Equals(Object o1, Object o2)
+        {
+            Assert.True(false, "Don\u2019t call this.");
```
```suggestion
            Assert.True(false, "Don't call this.");
```

AaronRobinsonMSFT commented on src/tests/Common/Assert.cs:
```diff
@@ -0,0 +1,240 @@
+
+// This file is used to provide some basic Assert functionality for assemblies that directly reference System.Private.CoreLib
+// and not the ref pack.
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
```
```suggestion
```

jkoritzinsky committed:
Update src/tests/Common/Assert.cs

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

jkoritzinsky committed:
Update src/tests/Common/Assert.cs

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

trylek reviewed (approved):
Looks great, thank you!

elinor-fung commented on src/tests/Interop/ReadMe.md:
```diff
@@ -10,7 +10,7 @@ There should be no more than **1** project type per folder (i.e. a folder can co
 Ancillary source assets for all tests should be located in `Interop/common` and can be easily added to all managed tests via the `Interop.settings.targets` file or native tests via `Interop.cmake`.
 
-A common pattern for testing is using the `Assert` utilities. This class is part of the `CoreCLRTestLibrary` which is included in all test projects by the `Interop.settings.targets` import. In order to use, add the following `using TestLibrary;` in the relevant test file.
+A common pattern for testing is using the `Assert` utilities. This class is part of the `CoreCLRTestLibrary` which is included in all test projects by the `Interop.settings.targets` import. In order to use, add the following `using Xunit;` in the relevant test file.
```
```suggestion
A common pattern for testing is using xUnit's `Assert` utilities. These utilities can be referenced via `CoreCLRTestLibrary` which is included in all test projects by `Directory.Build.targets` in this directory. In order to use, add the following `using Xunit;` in the relevant test file.
```

trylek replied:
Fixed in latest commit.

elinor-fung reviewed (approved):
Would it be useful to add a bullet about test utilities / asserts to https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/coreclr/test-configuration.md#adding-test-guidelines?

trylek committed:
Fix ThrowsWithInnerException and CollectionEqual

CollectionEqual ended up in infinite recursion as the array variant
no longer accepts the third string argument (there's no use for it
as Assert.Equal no longer supports it; after all, a single test was
using it) so that the enumerable variant ended up calling itself
instead of the array variant. I have removed the message string
and I deleted it from the one ResolveUnmanagedDllTests source.

In ThrowsWithInnerException there was a typo, there should be
"is TInner" in the inner exception check, not "is T".

Thanks

Tomas

trylek committed:
Address Elinor's PR feedback regarding interop readme doc

trylek commented:
The bugs are unrelated and known (https://github.com/dotnet/runtime/issues/61237, https://github.com/dotnet/runtime/issues/60152), merging in.

jkoritzinsky commented:
@tylek I think there's a conflicting change that got merged in before this one. Can you check the Pri0 test build and make sure it isn't broken? (I'm not near my computer otherwise I'd do it myself)

EgorBo commented:
Is failure like this https://dev.azure.com/dnceng/public/_build/results?buildId=1458028&view=logs&jobId=e2abd64e-6d96-553c-c262-deb1bd067c99&j=e2abd64e-6d96-553c-c262-deb1bd067c99&t=1849d1b9-84d9-551e-f982-c77c09278369 related?
`CoreCLR Common Pri0 Test Build AnyOS AnyCPU checked failed`
```
WeakReferenceTest.cs(260,20): error CS0117: 'Assert' does not contain a definition for 'IsNull' 
```
